[yyccbb] iP

yyccbb opened the pull request:
# Nihao
> "Say nihao to Nihao!"

Nihao frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/yyccbb/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :pilot:

And it is **FREE**!

### Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

***
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
     DataHandler.read();
     nihao.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

yyccbb committed:
Rename to Nihao. Create App and PrintHandler classes.

yyccbb committed:
Add echo functionality. Change personality.

yyccbb committed:
Add DataHandler class. Add temporary storage functionality.

yyccbb committed:
Create action, exception and handler packages. Refactor action firing logic.

yyccbb committed:
Add error handling. Add printed message for mark, unmark and list Actions.

yyccbb committed:
Add support for todo deadline and event. Refactor TaskAction.

yyccbb committed:
Configure batch automated text UI testing

yyccbb committed:
Create 3 exception classes. Update test input.

yyccbb committed:
Add delete task functionality. Update test cases accordingly.

yyccbb committed:
Use enums for commands

yyccbb committed:
DataHandler class: support local data persistence

DataHandler class does not support local data persistence. The list of
tasks created gets wiped away after one Nihao instance is closed.

Adding support for local data persistence allows users to save the tasks
so that they still can access them after closing the app.

Let's use JSON serialization and deserialisation functionalities to load
and save the list of tasks to a local JSON file.

JSON is used because it is possibly easier for integration with
databases.

yyccbb committed:
Merge branch 'branch-Level-7'

chonghaoen commented on src/main/java/handler/DataHandler.java:
```diff
@@ -0,0 +1,105 @@
+
+public class DataHandler {
+    private static final String fileNamePath = "data/PersistentData.json";
+    private static ArrayList<Task> tasks = new ArrayList<>();
```
Good naming separating single and multi-valued variables! 💯

chonghaoen commented on src/main/java/handler/PrintHandler.java:
```diff
@@ -0,0 +1,32 @@
+    public static void print(String msg) {
+        System.out.println(msg);
+    }
+    public static void printWithDivider(String msg) {
+        System.out.println(msg);
+        System.out.println(DIVIDER);
+    }
+    public static void printNumberedDivider(ArrayList<Task> msgs) {
```
Good and clear naming of methods! 💯

chonghaoen commented on src/main/java/util/TaskTypeAdapter.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Override
+    public Task read(JsonReader jsonReader) throws IOException {
+        Task ret = null;
```
Would there a better and clearer variable name for the task than ret? 🤔

chonghaoen commented on src/main/java/handler/InputHandler.java:
```diff
@@ -0,0 +1,109 @@
+        }
+    }
+
+    private int countByFlag(String[] parsedInput) {
+        int counter = 0;
+        for (int i = 0; i < parsedInput.length; i++) {
+            if (parsedInput[i].equals("/by")) {
+                counter++;
+            }
+        }
+        return counter;
+    }
+
+    private int countFromFlag(String[] parsedInput) {
+        int counter = 0;
+        for (int i = 0; i < parsedInput.length; i++) {
+            if (parsedInput[i].equals("/from")) {
+                counter++;
+            }
+        }
+        return counter;
+    }
+
+    private int countToFlag(String[] parsedInput) {
+        int counter = 0;
+        for (int i = 0; i < parsedInput.length; i++) {
+            if (parsedInput[i].equals("/to")) {
+                counter++;
+            }
+        }
+        return counter;
```
I like how clean this is! Would consider applying to my code too! 💯

yyccbb replied:
Thanks for the feedback!

chonghaoen reviewed (commented):
LGTM. Good code quality following naming standards.

yyccbb committed:
InputHandler class: remove own instance

yyccbb committed:
Task classes: store timestamps with LocalDateTime

Task classes store timestamps as String.

Storing timestamps as LocalDateTime objects is preferred for formatting
purposes and to prevent users from writing irrelevant strings as
timestamps.

Let's,
* create DateTimeHandler singleton class to handle conversion between
  String and LocalDateTime ojbects
* create IncorrectDateTimeFormatException class which produces
  exceptions to be thrown when a String that does not follow a given
  format attempts to get converted
* update Task classes to store LocalDateTime objects instead of Strings
* update Task classes' toString() methods to follow the specified format
* update TaskTypeAdapter to support LocalDateTime objects

yyccbb committed:
DeleteAction: support delete all

yyccbb committed:
Update textUI test

yyccbb committed:
Merge branch 'branch-Level-8'

yyccbb committed:
Create ExitAction class to handle "bye" command

"bye" command is handled directly by Nihao class.

This does not conform to the rest of the codebase where commands are
handled by InputHandler.

Let's create an ExitAction class to handle "bye" command.

yyccbb committed:
Create package nihao to host all packages and classes

yyccbb committed:
Merge branch 'add-gradle-support'

# Conflicts:
#	docs/README.md

yyccbb committed:
Configure build.gradle

yyccbb committed:
Configure codebase to use Gradle

yyccbb committed:
Action and Task classes: add .equals()

Action and Task classes do not have custom equals() methods.

The equals() method is needed for assertEquals() to assert equality
between action and task objects based on the needs of this application.

Let's create equals() methods for all Action and Task classes and create
InputHandlerTest class to test for InputHandler class.

yyccbb committed:
Create DataHandlerTest class

yyccbb committed:
DataHandler: create data directory when absent

yyccbb committed:
Add JavaDoc comments for most classes and methods

yyccbb committed:
Change some constants to variables

yyccbb committed:
TaskTypeAdapter: add class description

yyccbb committed:
Add find task by keyword functionality

'find' command is not supported.

It allows users to quickly search the list of tasks by a keyword.

Let's,
* create FindAction class to encapsulate this action
* update Command enum to include 'find' command
* update InputHandler to support 'find'
* update DataHandler to support this action
* update Task to add a method to return a boolean representing whether
  a substring is present in the task name or not

yyccbb committed:
Merge branch 'branch-A-JavaDoc'

yyccbb committed:
Merge branch 'branch-A-CodingStandard'

yyccbb committed:
Merge branch 'branch-Level-9'

yyccbb committed:
Update README.md

yyccbb committed:
Fix coding style

yyccbb committed:
Create GUI

Nihao does not have a GUI.

A GUI is more user-friendly.

Let's
* use JavaFX to create a GUI
* add start() method to Nihao as an entry point
* create DialogueBox class to generate reusable dialogue boxes
* update PrintHandler methods so that they return the content to be
  printed as String as well
* update Task.execute() so that they return the content to be printed as
  String, which is then packed up as dialogue boxes

yyccbb committed:
Update GUI to use fxml

yyccbb committed:
Change GUI theme to Cupertino-dark

yyccbb committed:
Merge branch 'branch-Level-10'

yyccbb committed:
Merge branch 'branch-A-CheckStyle'

yyccbb committed:
Use assertions in the code

Assertions are not used anywhere in the codebase.

They can be helpful in verifying the state of the program at strategic
points.

Let's,
* use assertions in the execute() methods of action classes to confirm
  that key fields have legal values when execute() is invocated
* use assertions in the getResponse() method of Nihao to confirm that
  the text input is not null

yyccbb committed:
DataHandler: refactor read() method

yyccbb committed:
DataHandler: refactor read() method

yyccbb committed:
DataHandler: improve code quality

yyccbb committed:
DataHandler: improve code quality

yyccbb committed:
PrintHandler: improve code quality

yyccbb committed:
PrintHandler: improve code quality

yyccbb committed:
Fix CheckStyle

yyccbb committed:
Fix CheckStyle

yyccbb committed:
Merge pull request #2 from yyccbb/branch-A-CodeQuality

Merge CodeQuality with master

yyccbb committed:
Revert "Merge CodeQuality with master"

yyccbb committed:
Merge pull request #3 from yyccbb/revert-2-branch-A-CodeQuality

Revert "Merge CodeQuality with master"

yyccbb committed:
Merge pull request #4 from yyccbb/branch-A-Assertions

Merge Assertions with master

yyccbb committed:
DataHandler: update method header comments

yyccbb committed:
DataHandler: update method header comments

yyccbb committed:
Merge pull request #5 from yyccbb/branch-A-CodeQuality

Merge CodeQuality with master

yyccbb committed:
Merge pull request #7 from yyccbb/new-branch-A-CodeQuality

Merge CodeQuality with master

yyccbb committed:
Add short aliases for all commands

yyccbb committed:
DialogueBox: fix text area width

yyccbb committed:
UI: add greeting message

yyccbb committed:
Update the user guide

yyccbb committed:
Update Gradle build configuration

yyccbb committed:
Fix a mistake in the user guide
